### PR TITLE
Add "satisfy respectively" assertion (#882)

### PR DESCRIPTION
**Pull request by vanashimko**
Add "satisfy respectively" assertion, similar to xUnit `Assert.Collection` (#882).

### CODE REVIEW

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 1]
```diff
@@ -349,6 +349,124 @@ private bool IsValidProperty<TSelector>(Expression<Func<T, TSelector>> propertyE
...
+                .ForCondition(Subject.Any())
+                .FailWith("but collection is empty.")
+                .Then
+                .ClearExpectation();
+
+            int elementsCount = Subject.Count();
+            int inspectorsCount = elementInspectors.Count;
+            if (elementsCount != inspectorsCount)
+            {
+                if (inspectorsCount == 1)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: I would simplify those two situations into a single generic statement:

`Expected {context:collection} to contain exactly {0} items, but it contained {1}  items"`

If the count doesn't match, it does not matter what kind of inspectors you used.

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 2]
```diff
@@ -349,6 +349,124 @@ private bool IsValidProperty<TSelector>(Expression<Func<T, TSelector>> propertyE
...
+        /// </param>
+        public void SatisfyRespectively(IEnumerable<Action<T>> expected, string because = "", params object[] becauseArgs)
+        {
+            if (expected is null)
+            {
+                throw new ArgumentNullException(nameof(expected), "Cannot verify against a <null> collection of inspectors");
+            }
+
+            ICollection<Action<T>> elementInspectors = expected.ConvertOrCastToCollection();
+            if (!elementInspectors.Any())
```
[the remark this answers is not in the document]
Comment (vanashimko) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: I thought about it and decided that passing empty collection of inspectors will be a mistake in 99% of cases 😄 But feel free to disagree.

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 3]
```diff
@@ -349,6 +349,124 @@ private bool IsValidProperty<TSelector>(Expression<Func<T, TSelector>> propertyE
...
 
+        /// <summary>
+        /// Asserts that a collection contains exactly a given number of elements, which meet
+        /// the criteria provided by the element inspectors.
+        /// </summary>
+        /// <param name="elementInspectors">
+        /// The element inspectors, which inspect each element in turn. The
+        /// total number of element inspectors must exactly match the number of elements in the collection.
+        /// </param>
+        public void SatisfyRespectively(params Action<T>[] elementInspectors)
```
[the remark this answers is not in the document]
Comment (vanashimko) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: We could, but after we verified each element and total count, what other things can we do? I've looked through the code (honestly speaking I've started with returning `AndConstraint`) and found that is such cases `void` is used (like in `BeEquivalentTo`).

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs (hunk 1 above)
Comment (vanashimko) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: Updated as suggested.

File: Src/FluentAssertions/Common/StringExtensions.cs [hunk 4]
```diff
@@ -86,6 +88,16 @@ public static string Capitalize(this string @this)
...
         }
 
+        /// <summary>
+        /// Appends tab character at the beginning of each line in a string.
+        /// </summary>
+        /// <param name="this"></param>
+        /// <returns></returns>
+        public static string IndentLines(this string @this)
+        {
+            return string.Join(Environment.NewLine,Regex.Split(@this, "\r\n|\r|\n").Select(x => $"\t{x}"));
```
[the remark this answers is not in the document]
Comment (vanashimko) [reply Src/FluentAssertions/Common/StringExtensions.cs]: Updated as suggested.

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jnyrup) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: If there's no reason to disallow chaining, I think we should allow it.
E.g. if the user want to call `BeInAscendingOrder`.

I'm aware of the `BeEquivalentTo` (and others) that returns a `void`, it's on my wish list for v6 to correct.

[the remark this answers is not in the document]
Comment (vanashimko) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: Agree, updated.

**Comment by vanashimko** [comment]
Hmm, tests failed two times in a row at AppVeyor, based on the tests failed, I suppose it's related to #1022. @jnyrup @dennisdoomen can I restart build somehow other than pushing new changes? Tests are passing stable on my machine, so fails are most likely not related to my changes.

**Comment by jnyrup** [comment]
@vanashimko Thanks for the contribution to Fluent Assertions.
Look out for your name in the next release.

**Comment by mdilovar** [comment]
Which version will this me available in?

**Comment by jnyrup** [comment]
@mdilovar It will be available in the next release, which will be 5.7.0.
I can't give a release date, but I can assure you that we're also eager to get the next release out.
We're working as I write on weeding out the last issues before being ready to push the release button.

**Pull request merged**